File: app/src/main/java/com/aticosoft/appointments/mobile/business/infrastructure/persistence/PersistenceModule.java
package com.aticosoft.appointments.mobile.business.infrastructure.persistence;

import com.aticosoft.appointments.mobile.business.infrastructure.persistence.PersistenceConfigurer.Services;

import javax.inject.Singleton;
import javax.jdo.PersistenceManagerFactory;

import dagger.Module;
import dagger.Provides;

/**
 * Created by rodrigo on 02/08/15.
 */
@Module
public class PersistenceModule {

    @Provides @Singleton
    protected PersistenceConfigurer providePersistenceConfigurer(Services services) {
        return new PersistenceConfigurer(services);
    }

    @Provides @Singleton
    protected PersistenceManagerFactory providePersistenceManagerFactory(PersistenceConfigurer persistenceConfigurer) {
        return persistenceConfigurer.createPersistenceManagerFactory();
    }
}

File: app/src/main/java/com/aticosoft/appointments/mobile/business/infrastructure/persistence/PersistenceModule.kt
package com.aticosoft.appointments.mobile.business.infrastructure.persistence

import com.aticosoft.appointments.mobile.business.ModulePostInitializer
import com.aticosoft.appointments.mobile.business.domain.model.common.persistable_object.TransactionalActionsManager
import com.aticosoft.appointments.mobile.business.domain.model.common.persistable_object.listener.async.PersistableObjectAsyncListenersManager
import com.aticosoft.appointments.mobile.business.domain.model.common.persistable_object.listener.sync.PersistableObjectSyncListenersManager
import com.aticosoft.appointments.mobile.business.infrastructure.persistence.configuration.PersistenceConfigurator
import com.aticosoft.appointments.mobile.business.infrastructure.persistence.configuration.PersistenceConfiguratorModule
import dagger.Module
import dagger.Provides
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Created by Rodrigo Quesada on 09/09/15.
 */
@Module(includes = arrayOf(PersistenceBaseModule::class, PersistenceConfiguratorModule::class))
/*internal*/ class PersistenceModule {

    @Singleton
    class PostInitializer @Inject protected constructor(
            private val objectSyncListenersManager: PersistableObjectSyncListenersManager,
            private val objectAsyncListenersManager: PersistableObjectAsyncListenersManager,
            private val transactionalActionsManager: TransactionalActionsManager
    ) : ModulePostInitializer {

        override fun init() {
            transactionalActionsManager.subscribeActions()
            objectSyncListenersManager.registerListeners()
            objectAsyncListenersManager.registerListeners()
        }
    }
}

@Module
/*internal*/ class PersistenceBaseModule {

    @Provides @Singleton
    fun providePersistenceManagerFactory(persistenceConfigurator: PersistenceConfigurator) = persistenceConfigurator.createPersistenceManagerFactory()
}